design/parallel-commit/parallel-commit-known-issues-and-solutions.md

```diff
@@ -0,0 +1,113 @@
+
+The memory lock should support range-based queries, because a reading operation may need to scan a range. So the memory lock should be implemented by an ordered map (like BTreeMap, Trie, SkipList).
+
+Therefore, the major difficulty of memory lock solution is that how to keep the performance.
+
+### Idea 2: RocksDB-Based Lock (By @gengliqi)
+
+This idea differs from [Idea1](#idea-1-key-based-memory-lock), only in that we are using RocksDB instead of an in-memory data structure to store locks. We avoid most performance impact if we can put it directly to LOCK_CF without Raft replication, however, it might introduce too many corner cases and troubles to resolve, considering that there are leader changing and region scheduling in TiKV. Therefore we may add a new CF to store these kind of locks. The performance comparing to [Idea1](#idea-1-key-based-memory-lock) is uncertain before we actually do a POC test.
+
+### Idea 3: Transaction List (By @Little-Wallace)
+
+For prewrite operations that need to exclude readings, add add them to a vector `V`. The `max_read_ts` should be acquired after the prewrite operation locks the latch of scheduler and adding itself to `V`.
+
+For point-get operations, access latch to see if the key is being prewritten with a smaller ts (prewrites of Parallel Commit transactions need to add its special information to the latch slot when it acquires the latch). If so, block or return KeyIsLocked err. For range scanning operations, check each items in `V` to see if their range overlaps. If so, block or return KeyIsLocked err.
+
+### Idea 4: Optimized Key-Based Memory Lock (By @Little-Wallace)
+
+The memlock consists of a lock-free ordered map like Idea 1 (for scanning request to check locks in range) and the latch of transaction scheduler like Idea 3 (for point getting requests).
+
+A prewrite of Parallel Commit transaction can get the `max_read_ts` first and then save the `max_read_ts` to the locks (both in the ordered map part and the latch) so that when a read operation checks locks, it can ignore the locks with `max_read_ts` greater than the current `read_ts`. However since this design is lock-free, it introduced another chance of breaking the isolation: a read may check lock between a prewrite getting `max_read_ts` and setting the memory lock. But this is quite easy to solve: let prewrite operation get another `max_read_ts` after locking the memlock. Thus the full procedure of prewrite looks like this:
+
+1. Atomically get the `max_read_ts` as `T1`.
+2. Lock the memlock and acquire latches, saving `T1` to them.
+3. Atomically get the `max_read_ts` again as `T2`.
+4. Continue performing prewrite and persist `T2` to the locks written down to engine.
+
+### Some other ideas
+
+* Make something like HLC, which doesn't seem to be practical for us.
+* ...
+
+## Replica Read (By @gengliqi)

```

Comment:
Another solution is `ReadIndex` carries read_ts and key range and treat it as normal read operations.